automation/Commands/utils/banner_utils.py

```diff
@@ -0,0 +1,129 @@
+
+    banner_list_url = "https://www.i-dont-care-about-cookies.eu/abp/"
+    urlretrieve(banner_list_url, path.join(output_directory, 'bannerlist.txt'))
+
+
+def parse_banner_list(banner_list_loc):
+    """Parse the banner selectors in the bannerlist and return them as a dictionary of lists.
+    """
+    filterlist = {}  # dictionary of {domain: list of banner selectors}
+
+    with open(banner_list_loc) as f:
+        for line in f:
+            if line.startswith('!'):
+                continue
+
+            # AdBlock's filter rules are complex, but we only care about a subset:
+            # element hiding rules.  These rules are denoted by the symbol "##".
+            # (see: https://adblockplus.org/en/filter-cheatsheet#elementhiding)
+            if "##" in line:
+                domains, selectors = line.strip().split("##")
+
+                # Commas denote a list of domains and selectors.
+                domains = [s.strip() for s in domains.split(",")]
+                selectors = [s.strip() for s in selectors.split(",")]
+
+                for domain in domains:
+                    assert not domain.startswith('~')  # currently unsupported
+                    domain = '__global__' if domain == '' else domain
+                    filterlist.setdefault(domain, []).extend(selectors)
+            else:
+                # the rest appear to be rules to block scripts
+                filterlist.setdefault('__unknown__', []).append(line.strip())
+
+    return filterlist
+
+
+def find_banners_by_selectors(page_url, page_html, banner_list_loc, logger=None):
+    """ Finds cookie-notice/cookie-wall banners on an HTML page using CSS selectors. The
+    selectors are collected by the browser extension "I don't care about cookies".
+
+    Args:
+        page_url: URL of the webpage, used to add domain-specific selectors
+        page_html: HTML of the webpage
+        banner_list_loc: location of selector file (ABP filter-list style)
+        logger: optional paramter to report progress; if None, stdout is used
+    Returns:
+        Tuple: (number of selectors, list of banner elements)
+        The banners are in a named tuple (selector, element-tag, element-id, banner-text)
+
+    Notes:
+        Since the selectors are collected using crowd-sourcing, and thus might miss sites.
+        There are additional false-negatives and false-positive problems as well, but the
+        overall  accuracy is high for well-known sites (>90%).
+
+        The false-negatives are caused by banners created using javascript redirects, which
+        the extension detects, but are not in their ABP filter list (e.g. see dumpert.nl).
+        The false-positives are multiple matches (e.g. nu.nl) or matches to invisible elements
+        (e.g. faz.net).
+
+        The speed of this function varies based on the passed HTML -- from a few seconds to
+        almost a minute.
+    """
+    time0 = time()
+    log_func = logger.info if logger else print
+
```

Comment:
I added here some print-outs:

```
    log_func("ROBERT DEBUG BEGIN")
    log_func(page_url)
    log_func(page_html)
    log_func(banner_list_loc)
    log_func("ROBERT DEBUG END")
```

So I noticed that sometimes `banner_list_loc` is only `/tmp` instead of a file.